src/celeritas/neutron/model/NeutronInelasticModel.cc

```diff
@@ -0,0 +1,237 @@
+//---------------------------------------------------------------------------//
+//! \file celeritas/neutron/model/NeutronInelasticModel.cc
+//---------------------------------------------------------------------------//
+#include "NeutronInelasticModel.hh"
+
+#include "corecel/math/ArrayOperators.hh"
+#include "corecel/math/Quantity.hh"
+#include "celeritas/global/CoreParams.hh"
+#include "celeritas/global/CoreState.hh"
+#include "celeritas/io/ImportPhysicsVector.hh"
+#include "celeritas/mat/MaterialParams.hh"
+#include "celeritas/phys/InteractionApplier.hh"
+#include "celeritas/phys/PDGNumber.hh"
+#include "celeritas/phys/ParticleParams.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from model ID and other necessary data.
+ */
+NeutronInelasticModel::NeutronInelasticModel(ActionId id,
+                                             ParticleParams const& particles,
+                                             MaterialParams const& materials,
+                                             ReadData load_data)
+{
+    CELER_EXPECT(id);
+    CELER_EXPECT(load_data);
+
+    HostVal<NeutronInelasticData> data;
+
+    // Save IDs
+    data.scalars.action_id = id;
+    data.scalars.neutron_id = particles.find(pdg::neutron());
+
+    CELER_VALIDATE(data.scalars.neutron_id,
+                   << "missing neutron particles (required for "
+                   << this->description() << ")");
+
+    // Save particle properties
+    data.scalars.neutron_mass = particles.get(data.scalars.neutron_id).mass();
+
+    // Load neutron inelastic cross section data
+    make_builder(&data.micro_xs).reserve(materials.num_elements());
+    for (auto el_id : range(ElementId{materials.num_elements()}))
+    {
+        AtomicNumber z = materials.get(el_id).atomic_number();
+        this->append_micro_xs(load_data(z), &data);
+    }
+    CELER_ASSERT(data.micro_xs.size() == materials.num_elements());
+
+    // Build nucleon-nucleon cross section data
+    size_type num_channels = data.scalars.num_channels();
+    make_builder(&data.nucleon_xs).reserve(num_channels);
+    auto xs_params = make_builder(&data.xs_params);
+    xs_params.reserve(num_channels);
+
+    auto bins = this->get_channel_bins();
+    for (auto channel_id : range(ChannelId{num_channels}))
+    {
+        // Add nucleon-nucleon cross section parameters and data
+        ChannelXsData channel_data = this->get_channel_xs(channel_id);
+        xs_params.push_back(channel_data.par);
+
+        auto reals = make_builder(&data.reals);
+        GenericGridData nucleon_xs;
+
+        auto xs = make_span(channel_data.xs);
+        CELER_ASSERT(xs.size() == bins.size());
+        nucleon_xs.grid = reals.insert_back(bins.begin(), bins.end());
+        nucleon_xs.value = reals.insert_back(xs.begin(), xs.end());
+        CELER_ASSERT(nucleon_xs);
+        make_builder(&data.nucleon_xs).push_back(nucleon_xs);
```

Comment:
Same here, this could use the `GenericGridBuilder`:
```suggestion
        GenericGridBuilder build_grid{&data.reals};
        make_builder(&data.nucleon_xs)
            .push_back(build_grid(bins, make_span(channel_data.xs)));
```

Reply:
Done.